### PR TITLE
feat: unify setup-dev flow into shared setup module

### PR DESCRIPTION
**Pull request by nick-inkeep**
## Summary

Extracts the setup-dev flow into a shared `@inkeep/agents-core/setup` module. The monorepo's bash `setup.sh` (223 lines) and the template's `setup.js` (623 lines) are replaced by thin wrappers (~30-60 lines) that pass config to a single `runSetup()` entry point.

### Why a shared module in `agents-core` (not a standalone package or script)

The setup flow needs access to `loadEnvironmentFiles()` from `agents-core` and must be importable by both the monorepo (`scripts/setup-dev.js`) and the published quickstart template (`create-agents-template`). Putting it in `agents-core/setup` avoids a new package while keeping it co-located with the env/auth/db primitives it depends on. Trade-off: `agents-core` grows slightly, and the monorepo wrapper must build `agents-core` before running setup (visible in the Cypress CI change).

### Config-driven callers instead of flags

Each caller passes a `SetupConfig` object (compose file path, migration commands, push target, server commands, `isCloud`, `skipPush`). This avoids a flag explosion on a single CLI and lets each context (monorepo contributor vs. quickstart user vs. cloud deploy) declare exactly what it needs. The shared module has zero awareness of who's calling it.

### Pre-flight checks before Docker startup

Rather than letting `docker-compose up` fail cryptically on port conflicts, the module runs `lsof` + `docker ps` + `docker inspect` to identify which Docker Compose project owns the conflicting port and tells the user the exact `docker compose -p <name> down` command. For standalone (non-compose) containers, it suggests `docker rm -f <name>`. If the ports belong to our own project, startup is skipped (idempotent re-run).

When Docker is unavailable, the module checks whether database URLs point to localhost or an external host. Localhost URLs (from `.env.example` defaults) indicate Docker-managed DBs are expected, so setup exits with an error. External host URLs indicate the user is managing DBs externally, so Docker is skipped gracefully.

### Health polling: all three database services

After `docker-compose up -d`, health polling checks all three services in parallel: `doltgres-db` (60s timeout), `postgres-db` (30s), and `spicedb-postgres` (30s). Setup only hard-fails if both primary databases (doltgres + postgres) fail health checks — a SpiceDB-only failure logs a warning but continues, since `writeSpiceDbSchema()` in auth init has its own 30-retry logic.

### Auth strategy: bypass secret vs. CLI keychain

When `INKEEP_AGENTS_MANAGE_API_BYPASS_SECRET` is available, push uses CI mode (`INKEEP_CI=true` + `INKEEP_API_KEY`). When it's absent (cloud users), the CLI resolves auth from keychain/profile — the template pre-runs `inkeep init --no-interactive` + `inkeep login` before calling `runSetup()`. This means local contributors and cloud users follow different auth paths through the same code, controlled entirely by the presence of the bypass secret.

### Process group cleanup for temporarily spawned servers

When the API server is spawned for project push, it's started with `detached: true` so it gets its own process group. Cleanup uses `kill(-pid)` (negative PID = group signal) followed by a 2-second grace period and `SIGKILL`. This prevents orphaned Node processes after setup completes.

### CI concurrency limit

Added `--concurrency=8` to `pnpm check` in CI. On cache-miss runs, unlimited Turbo concurrency causes multiple Next.js builds to run simultaneously on a 4-vCPU runner, triggering OOM kills. This is tangential to the setup unification but was discovered and fixed during CI validation of this PR.

## User scenarios addressed

| Scenario | Behavior |
|---|---|
| First-time contributor runs `pnpm setup-dev` | Full flow: .env creation, JWT gen, Docker startup, migrations, auth init, project push |
| Re-run after setup already completed | Idempotent: skips JWT gen (keys exist), skips Docker startup (containers running), migrations apply only what's new |
| Another Docker project occupies port 5432/5433/5434 | Pre-flight detects conflict, reports project name and exact `docker compose -p <name> down` command (or `docker rm -f <name>` for standalone containers), continues with existing DBs |
| Docker not installed, DB URLs point to external host | Graceful degradation: skips Docker entirely, runs migrations against external DBs |
| Docker not installed, DB URLs are localhost | Exits with error — localhost URLs indicate Docker-managed DBs are expected |
| `pnpm setup-dev --skip-push` | Runs env/Docker/migrations/auth but skips server spawn and project push |
| `docker compose up` slow on cold start | Timeout (120s) triggers warning, falls through to health polling which waits for containers to become healthy |
| Quickstart template setup (local) | Same flow via `@inkeep/agents-core/setup`, different compose file and migration commands |
| Quickstart template setup (cloud, no bypass secret) | Runs `inkeep init` + `inkeep login` before setup, push authenticates via CLI keychain |
| Cypress E2E in CI | Builds `agents-core` first, runs `setup-dev --skip-push` (no project push needed for E2E) |

## Changes

- **`packages/agents-core/src/setup/setup.ts`** — Shared setup module (~730 lines). Includes `isLocalDatabaseUrl()` helper for localhost detection and health polling for all 3 DB services.
- **`scripts/setup-dev.js`** — Monorepo wrapper (32 lines), replaces deleted `scripts/setup.sh`
- **`create-agents-template/scripts/setup.js`** — Rewritten as thin wrapper
- **`.github/workflows/ci.yml`** — `--concurrency=8` on `pnpm check`
- **`.github/composite-actions/cypress-e2e/action.yml`** — Build `agents-core` before setup; `--skip-push`
- **`agents-cookbook/template-projects/inkeep.config.ts`** — Added `apiKey` from bypass secret env var
- **`agents-docs/`** — Updated contributing overview + added troubleshooting section for local env issues
- **`.gitignore`** — Added `.setup-complete` marker file

### Unit tests (17 cases in `setup.test.ts`)

Covers the core `runSetup()` logic with mocked `exec`/`spawn`/`fetch`:
- Migration execution (manage + runtime, parallel)
- Auth init (with credentials, without credentials)
- Docker skip when `isCloud: true`
- Docker startup when ports are free
- SpiceDB PostgreSQL health-checked alongside doltgres and postgres
- Docker skip when pre-flight detects foreign containers on required ports
- Docker compose timeout during startup (graceful continue to health polling)
- Docker unavailable with external DB URLs (graceful skip)
- Docker unavailable with localhost DB URLs (exits with error)
- Missing DB URLs on non-cloud (exits with error)
- Project push command format and env var pass-through
- Push with bypass secret (CI mode: `INKEEP_CI`, `INKEEP_API_KEY`, `INKEEP_TENANT_ID`)
- Push without bypass secret (CLI keychain auth, no CI env vars)
- `--skip-push` flag honored
- Server already running (no spawn)
- Server not running (spawn + wait for health)

## Test plan

_Manual QA scenarios tested on macOS (2026-02-18). These resist automation because they require real Docker, real database startups, and real process lifecycle behavior._

- [x] **Clean-state full flow**: `pnpm setup-dev` from scratch (after `docker-compose down -v` + removing `.setup-complete`). All 8 steps completed: .env check, JWT skip (already configured), Docker startup with health polling, manage + runtime migrations, auth init, API server temporary spawn, weather-project push, server cleanup. · _Why not a test: requires real Docker daemon, real database containers, real network health checks._
- [x] **Idempotency re-run**: `pnpm setup-dev` immediately after successful first run. JWT gen skipped ("already configured"), Docker `up -d` is a no-op (containers already healthy), all three health checks pass (DoltgreSQL, PostgreSQL, SpiceDB PostgreSQL), migrations ran cleanly, project re-pushed successfully. · _Why not a test: requires persistent state across runs (`.setup-complete` file, running containers)._
- [x] **`--skip-push` flag**: `pnpm setup-dev --skip-push` completes after Step 5 (auth init). No Step 6/7/8 output, no API server spawned, ends with "Database setup completed!" message. · _Why not a test: verifiable via unit test (already covered), but manual run confirms real end-to-end flag plumbing._
- [x] **Port conflict detection**: Stopped our containers, started a foreign postgres on port 5432 (`docker run -d --name conflict-test -p 5432:5432 postgres:16`). Setup correctly detected `:5432 (DoltgreSQL) → conflict-test`, printed the conflict warning, skipped Docker startup, and proceeded to migrations. · _Why not a test: requires real Docker containers, real port binding, real `lsof` + `docker inspect` calls._
- [x] **Localhost detection regex**: Validated `isLocalDatabaseUrl()` against 10 edge cases including false-positive guards (localhost in DB name, localhost in password). · _Why not a test: could be a unit test; validated via inline Node script during QA._
- [x] **CI pipeline**: build, typecheck, lint, unit tests all green.
- [x] **Create Agents E2E**: Template setup flow green.
- [x] **Cypress E2E**: Docker-missing graceful degradation path green.

### Bugs found and fixed during QA

1. **Exec timeout misclassified as port conflict** (fixed in `cee4e1eb2`): `startDockerDatabases()` treated `killed === true` (Node exec timeout) as a port conflict, skipping health polling. On cold starts where Docker needed to pull images, this caused migrations to fail against containers that weren't ready. Fix: separated timeout from port-conflict handling — timeouts now fall through to health polling, and the exec timeout was increased from 60s to 120s.

2. **No fix suggestion for non-compose containers** (fixed in `cee4e1eb2`): Port conflict detection showed an empty "Fix:" section for standalone containers without a `com.docker.compose.project` label. Fix: added `docker rm -f <name>` fallback for non-compose containers.

### Review feedback addressed

3. **DB URL existence check was meaningless** (fixed in `22d5934`): `.env.example` always provides localhost DB URLs, so the "Docker not available but DB URLs are set" fallback always passed. Fix: replaced existence check with `isLocalDatabaseUrl()` — localhost URLs require Docker; external host URLs allow graceful Docker skip.

4. **SpiceDB PostgreSQL not health-checked** (fixed in `22d5934`): Health polling only checked `doltgres-db` and `postgres-db`, missing `spicedb-postgres` (port 5434). Fix: added `spicedb-postgres` to the `Promise.allSettled` health polling with 30s timeout.

## Future considerations

- **`waitForServerReady` exists in 4 variants** across the repo (`setup.ts`, `create-agents/e2e/utils.ts`, `agents-cli/dev.ts`, `agents-sdk/module-hosted-tool-manager.ts`). Consolidation was considered but deferred — the implementations differ intentionally (exponential backoff vs. fixed interval, CI-aware logging vs. silent, HTML body check vs. status code, etc.). A unified function flexible enough for all cases would be more complex than the ~15-line inline versions. Revisit if more call sites emerge.

---

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: packages/agents-core/src/setup/setup.ts [hunk 1]
```diff
@@ -0,0 +1,542 @@
...
+      /* already dead */
+    }
+    try {
+      process.kill(pid, 'SIGKILL');
+    } catch {
+      /* already dead */
+    }
+  }, 2000);
+  logSuccess(`${name} stopped`);
+}
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/setup/setup.ts:610]: **Declined.** The entire setup module is Unix-only throughout: `spawn('sh', ['-c', ...])`, `docker-compose` via shell strings, SIGTERM/SIGKILL process groups. Windows would fail at line 344 (`sh` shell), line 142 (`docker-compose`), and line 159 (`docker inspect`) long before reaching `stopProcessGroup`. Adding `taskkill` conditionals would be dead code on a tool with Unix-only prerequisites. If Windows support becomes a goal, it would need a holistic effort, not a point fix.

File: packages/agents-core/src/setup/setup.ts [hunk 2]
```diff
@@ -0,0 +1,542 @@
...
+
+    if (!privateKeyFound) lines.push(`INKEEP_AGENTS_TEMP_JWT_PRIVATE_KEY=${privateKeyBase64}`);
+    if (!publicKeyFound) lines.push(`INKEEP_AGENTS_TEMP_JWT_PUBLIC_KEY=${publicKeyBase64}`);
+
+    await writeFile('.env', lines.join('\n'));
+    logSuccess('JWT keys generated and added to .env');
+  } catch {
+    logWarning('Failed to generate JWT keys - playground may not work');
+    logInfo('You can manually run: pnpm run generate-jwt-keys');
+  }
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/setup/setup.ts]: **Accepted.** Good catch — `logError(msg, error)` is the established pattern throughout this file (e.g., `pushProject` at line 447, `startDockerDatabases` at line 210). Changed `catch` to `catch (error)` and `logWarning` to `logError` to surface the actual error message. Fixed in next push.

File: packages/agents-core/src/setup/setup.ts [hunk 3]
```diff
@@ -0,0 +1,542 @@
...
+    const proc = spawn('sh', ['-c', config.devApiCommand], {
+      stdio: 'ignore',
+      detached: true,
+      cwd: process.cwd(),
+    });
+    proc.unref();
+    result.startedApi = true;
+    result.apiPid = proc.pid ?? null;
+    logSuccess(`API server process started (PID: ${proc.pid})`);
+  }
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/setup/setup.ts:539]: **Declined — valid but out of scope.** The issue is real: if the spawn fails immediately, users wait for the 180s health-check timeout. However, the health-check timeout does surface the failure (line 388: `logWarning('Server readiness check failed: ...')`), so it's not a silent failure.

The suggested fix requires switching from `detached+unref+stdio:ignore` to capturing stderr and managing process lifecycle differently — a meaningful behavioral change. This PR is focused on unifying the setup flow with zero code duplication, not rewriting the server lifecycle. I'd prefer this as a follow-up improvement.

File: packages/agents-core/src/setup/setup.ts [hunk 4]
```diff
@@ -0,0 +1,542 @@
...
+}
+
+const SETUP_COMPLETE_FILE = '.setup-complete';
+
+async function ensureEnvFile() {
+  if (!existsSync('.env') && existsSync('.env.example')) {
+    copyFileSync('.env.example', '.env');
+    logSuccess('Created .env from .env.example');
+  }
+}
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply packages/agents-core/src/setup/setup.ts:266]: **Declined.** The function gates on `existsSync('.env.example')` before calling `copyFileSync`, so the source is confirmed to exist. The destination is confirmed not to exist. The remaining failure scenarios (write permission denied, disk full) are narrow edge cases where:

1. An unhandled exception propagating to the caller is the correct behavior — setup genuinely cannot proceed without a `.env` file
2. Wrapping in try/catch with `process.exit(1)` achieves the same outcome (crash) but with more code
3. The raw error from `copyFileSync` already contains the diagnostic information (EACCES, ENOSPC, etc.)

Adding a try/catch here would add noise without improving the outcome.

File: scripts/setup-dev.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env node
```
[the remark this answers is not in the document]
Comment (nick-inkeep) [reply scripts/setup-dev.js:1]: **Declined.** Root `package.json` has `"type": "module"`, so `.js` files are already treated as ESM — the top-level `await` and bare `import` in this file confirm it works correctly. The `.mjs` files in `scripts/` predate the `"type": "module"` setting. Renaming to `.mjs` would require updating the `package.json` script reference for zero functional benefit — purely cosmetic churn in a PR focused on unifying setup logic.

File: packages/agents-core/src/setup/setup.ts [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+  // Check 1: Docker daemon is running
+  try {
+    await execAsync('docker info', { timeout: 10000 });
+  } catch {
+    if (
+      process.env.INKEEP_AGENTS_MANAGE_DATABASE_URL &&
+      process.env.INKEEP_AGENTS_RUN_DATABASE_URL
+    ) {
+      logWarning('Docker is not available, but database URLs are configured');
+      logInfo('Assuming databases are managed externally.');
```
Comment (miles-kt-inkeep) [inline packages/agents-core/src/setup/setup.ts:131]: The default environment variables from .env.example get copied into .env before this runs, so I don't think this assumption makes sense. You could check for `localhost` or something in the urls to determine if they are managed by docker

File: packages/agents-core/src/setup/setup.ts [hunk 7]
```diff
@@ -0,0 +1,720 @@
...
+      return;
+    } else {
+      logError('Failed to start database containers', error);
+      process.exit(1);
+    }
+  }
+
+  logInfo('Polling Docker health status...');
+
+  const [doltgresResult, postgresResult] = await Promise.allSettled([
```
Comment (miles-kt-inkeep) [inline packages/agents-core/src/setup/setup.ts]: do we need to check spice db also?

File: packages/agents-core/src/setup/setup.ts [hunk 8]
```diff
@@ -0,0 +1,720 @@
...
+async function checkServerRunning(url: string, timeout = 5000): Promise<boolean> {
+  try {
+    const response = await fetch(url, { signal: AbortSignal.timeout(timeout) });
+    return response.ok || response.status === 204;
+  } catch {
+    return false;
+  }
+}
+
+async function waitForServerReady(url: string, timeout = 180000) {
```
Comment (miles-kt-inkeep) [inline packages/agents-core/src/setup/setup.ts:485]: This is a duplicate functionality to the helper in `packages/create-agents/src/__tests__/e2e/utils.ts‎` maybe consolidate?

**Review by miles-kt-inkeep — APPROVED** [review]
